src/Microsoft.VisualStudio.Composition/ByValueEquality.cs

```diff
@@ -13,17 +13,19 @@ namespace Microsoft.VisualStudio.Composition
 
     internal static partial class ByValueEquality
     {
-        internal static IEqualityComparer<IReadOnlyDictionary<string, object>> Metadata
+        internal static IEqualityComparer<IReadOnlyDictionary<string, object?>> Metadata
         {
             get { return MetadataDictionaryEqualityComparer.Default; }
         }
 
-        internal static IEqualityComparer<IReadOnlyDictionary<TKey, TValue>> Dictionary<TKey, TValue>(IEqualityComparer<TValue> valueComparer = null)
+        internal static IEqualityComparer<IReadOnlyDictionary<TKey, TValue>> Dictionary<TKey, TValue>(IEqualityComparer<TValue>? valueComparer = null)
+            where TKey : notnull
         {
             return DictionaryEqualityComparer<TKey, TValue>.Get(valueComparer);
         }
 
         internal static IEqualityComparer<IReadOnlyDictionary<TKey, ImmutableHashSet<TValue>>> DictionaryOfImmutableHashSet<TKey, TValue>()
+            where TKey : notnull
```

Comment:
I needed something just like this, and was unable to find a solution that didn't limit to reference types (but it clearly exists) - where is `notnull` coming from?

Reply:
Nevermind, I found it in the docs. But I'm not sure why it wasn't available for me in the VSLSP repo - maybe I just wasn't using the right constraint... anyway ignore me :smile: